In [1]:
import pandas as pd
import numpy as np

In [2]:
#Importing DF
#The Data
# https://www.kaggle.com/uciml/mushroom-classification 
Shroom_DF = pd.read_csv('mushroom-classification/mushrooms.csv', sep=',')

In [3]:
#Exploring Data
Shroom_DF.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [4]:
#Checking to see what our data contains 
Shroom_DF.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [5]:
#Checkin to see if we are missing any data
Shroom_DF.isnull().any()

class                       False
cap-shape                   False
cap-surface                 False
cap-color                   False
bruises                     False
odor                        False
gill-attachment             False
gill-spacing                False
gill-size                   False
gill-color                  False
stalk-shape                 False
stalk-root                  False
stalk-surface-above-ring    False
stalk-surface-below-ring    False
stalk-color-above-ring      False
stalk-color-below-ring      False
veil-type                   False
veil-color                  False
ring-number                 False
ring-type                   False
spore-print-color           False
population                  False
habitat                     False
dtype: bool

In [6]:
#Have to convert categorical data into int form to use for classifiers.
Dum_Shroom_DF = pd.get_dummies(Shroom_DF, drop_first= True)

In [7]:
Dum_Shroom_DF.head()

class_p  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  cap-shape_x  \
0        1            0            0            0            0            1   
1        0            0            0            0            0            1   
2        0            0            0            0            0            0   
3        1            0            0            0            0            1   
4        0            0            0            0            0            1   

   cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  ...  \
0              0              1              0            0  ...   
1              0              1              0            0  ...   
2              0              1              0            0  ...   
3              0              0              1            0  ...   
4              0              1              0            0  ...   

   population_n  population_s  population_v  population_y  habitat_g  \
0             0             1             0             0          0   
1             1             0             0             0          1   
2             1             0             0             0          0   
3             0             1             0             0          0   
4             0             0             0             0          1   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0          0          0          0          1          0  
1          0          0          0          0          0  
2          0          1          0          0          0  
3          0          0          0          1          0  
4          0          0          0          0          0  

[5 rows x 96 columns]

In [8]:
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

In [9]:
import matplotlib.pyplot as plt

In [10]:
#Creating our features and labels
Shroom_Features = Dum_Shroom_DF.loc[:, Dum_Shroom_DF.columns !='class_p'].copy()
Shroom_Label = Dum_Shroom_DF[['class_p']].copy()


In [11]:
Shroom_Features.head()

cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  cap-shape_x  \
0            0            0            0            0            1   
1            0            0            0            0            1   
2            0            0            0            0            0   
3            0            0            0            0            1   
4            0            0            0            0            1   

   cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  cap-color_e  ...  \
0              0              1              0            0            0  ...   
1              0              1              0            0            0  ...   
2              0              1              0            0            0  ...   
3              0              0              1            0            0  ...   
4              0              1              0            0            0  ...   

   population_n  population_s  population_v  population_y  habitat_g  \
0             0             1             0             0          0   
1             1             0             0             0          1   
2             1             0             0             0          0   
3             0             1             0             0          0   
4             0             0             0             0          1   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0          0          0          0          1          0  
1          0          0          0          0          0  
2          0          1          0          0          0  
3          0          0          0          1          0  
4          0          0          0          0          0  

[5 rows x 95 columns]

In [12]:
Shroom_Label.head()

class_p
0        1
1        0
2        0
3        1
4        0

In [13]:
#Setting up the Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier
Rand_For_Class = RandomForestClassifier(n_estimators = 10, random_state = 15)
X = Shroom_Features
y= Shroom_Label
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10)
Rand_For_Class.fit(X_train, y_train.values.ravel())


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=15, verbose=0, warm_start=False)

In [14]:
#Accuracy
Forest_Pred = Rand_For_Class.predict(X_test)
target_names = ["Edible", 'Poisonous']

print(classification_report(y_test, Forest_Pred,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_test, Forest_Pred)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1276
   Poisonous       1.00      1.00      1.00      1162

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy Score 1.0


Next Classifier KNN

In [15]:
from sklearn.neighbors import KNeighborsClassifier

In [16]:
#Test is going to be split same way as above, just reposting the same code
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=15)


In [17]:
knn = KNeighborsClassifier(n_neighbors = 5, metric = 'euclidean')
knn.fit(X_train, y_train.values.ravel())

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [18]:
Knn_Pred = knn.predict(X_test)

In [19]:
target_names = ["Edible", 'Poisonous']

print(classification_report(y_test, Knn_Pred,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_test, Knn_Pred)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1263
   Poisonous       1.00      1.00      1.00      1175

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy Score 1.0


In [20]:
#KNN as 10
knn = KNeighborsClassifier(n_neighbors = 10, metric = 'euclidean')
knn.fit(X_train, y_train.values.ravel())
Knn_Pred = knn.predict(X_test)
print(classification_report(y_test, Knn_Pred,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_test, Knn_Pred)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1263
   Poisonous       1.00      1.00      1.00      1175

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy Score 1.0


In [21]:
#KNN as 20
knn = KNeighborsClassifier(n_neighbors = 20, metric = 'euclidean')
knn.fit(X_train, y_train.values.ravel())
Knn_Pred = knn.predict(X_test)
print(classification_report(y_test, Knn_Pred,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_test, Knn_Pred)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1263
   Poisonous       1.00      1.00      1.00      1175

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy Score 0.9991796554552912


In [22]:
#KNN as 40
knn = KNeighborsClassifier(n_neighbors = 40, metric = 'euclidean')
knn.fit(X_train, y_train.values.ravel())
Knn_Pred = knn.predict(X_test)
print(classification_report(y_test, Knn_Pred,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_test, Knn_Pred)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1263
   Poisonous       1.00      1.00      1.00      1175

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy Score 0.9983593109105825


In [23]:
# SVM Classifier 

from sklearn.svm import SVC
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=15)

svclassifier = SVC(kernel = 'linear')
svclassifier.fit(X_train, y_train.values.ravel())


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='linear', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [24]:
SVC_pred = svclassifier.predict(X_test)
target_names = ['Edible', 'Poisonous']
print(classification_report(y_test, SVC_pred, target_names=target_names))
print("Accuracy:",metrics.accuracy_score(y_test, SVC_pred))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1263
   Poisonous       1.00      1.00      1.00      1175

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy: 1.0


In [25]:
from sklearn.tree import DecisionTreeClassifier

In [26]:
#Decision Tree Classifier
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=15)
DTC = DecisionTreeClassifier(random_state = 5)
DTC.fit(X_train, y_train.values.ravel())                                                  

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=5,
            splitter='best')

In [27]:
DTC_pred = DTC.predict(X_test)
target_names = ['Edible', 'Poisonous']
print(classification_report(y_test, DTC_pred, target_names=target_names))
print("Accuracy:",metrics.accuracy_score(y_test, DTC_pred))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00      1263
   Poisonous       1.00      1.00      1.00      1175

   micro avg       1.00      1.00      1.00      2438
   macro avg       1.00      1.00      1.00      2438
weighted avg       1.00      1.00      1.00      2438

Accuracy: 1.0


After an initial run just to see how the sets would run, it looks very positive.  Now I will use a validation set to look for official results.

In [28]:
from sklearn.metrics import confusion_matrix

In [29]:
#Validation Set of .1

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.125, random_state=15)

# .7 Training / .1 val / .2 Test
# .8 of the original train * .125 will equal .1  Subtract .8 with .1 because .1 is the validation set and .7 is the train
#leaving the remaining .2 as the test set

In [30]:
Rand_For_Class.fit(X_train, y_train.values.ravel())

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=15, verbose=0, warm_start=False)

In [31]:
#Accuracy
Forest_Pred = Rand_For_Class.predict(X_val)
target_names = ["Edible", 'Poisonous']

print(classification_report(y_val, Forest_Pred,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Forest_Pred)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy Score 1.0


In [32]:
tn, fp, fn, tp = confusion_matrix(y_val, Forest_Pred).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


426 Edible mushrooms were labeled correctly and 387 poisonous mushrooms were labeled correctly.  

In [33]:
knn_LF = KNeighborsClassifier(n_neighbors = 5, metric = 'euclidean')
knn_LF.fit(X_train, y_train.values.ravel())
Knn_Pred_LF = knn_LF.predict(X_val)
print(classification_report(y_val, Knn_Pred_LF,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_LF)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy Score 1.0


In [34]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_LF).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


Another perfect classification!

In [35]:
knn_LF = KNeighborsClassifier(n_neighbors = 10, metric = 'euclidean')
knn_LF.fit(X_train, y_train.values.ravel())
Knn_Pred_LF2 = knn_LF.predict(X_val)
print(classification_report(y_val, Knn_Pred_LF2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_LF2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy Score 1.0


In [36]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_LF2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


In [37]:
knn_LF = KNeighborsClassifier(n_neighbors = 20, metric = 'euclidean')
knn_LF.fit(X_train, y_train.values.ravel())
Knn_Pred_LF3 = knn_LF.predict(X_val)
print(classification_report(y_val, Knn_Pred_LF3,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_LF3)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy Score 1.0


In [38]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_LF3).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


In [39]:
knn_LF = KNeighborsClassifier(n_neighbors = 40, metric = 'euclidean')
knn_LF.fit(X_train, y_train.values.ravel())
Knn_Pred_LF4 = knn_LF.predict(X_val)
print(classification_report(y_val, Knn_Pred_LF4,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_LF4)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy Score 1.0


In [40]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_LF4).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


In [41]:
#SVM
svclassifier_LF = SVC(kernel = 'linear')
svclassifier_LF.fit(X_train, y_train.values.ravel())

SVC_pred_LF = svclassifier_LF.predict(X_val)

print(classification_report(y_val, SVC_pred_LF, target_names=target_names))
print("Accuracy:",metrics.accuracy_score(y_val, SVC_pred_LF))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy: 1.0


In [42]:
tn, fp, fn, tp = confusion_matrix(y_val, SVC_pred_LF).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


SVC does as good of a job as the rest

In [43]:
DTC_LF = DecisionTreeClassifier(random_state = 15)
DTC_LF.fit(X_train, y_train.values.ravel()) 
DTC_pred_LF = DTC_LF.predict(X_val)

print(classification_report(y_val, DTC_pred_LF, target_names=target_names))
print("Accuracy:",metrics.accuracy_score(y_val, DTC_pred_LF))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       426
   Poisonous       1.00      1.00      1.00       387

   micro avg       1.00      1.00      1.00       813
   macro avg       1.00      1.00      1.00       813
weighted avg       1.00      1.00      1.00       813

Accuracy: 1.0


In [44]:
tn, fp, fn, tp = confusion_matrix(y_val, DTC_pred_LF).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  426
False Positives:  0
False Negatives:  0
True Positives:  387


Looks like the decision tree classifier works as well as the others

Sadly, the amount of data we have to play with is not that much, meaning our validation set results may be just that good or it is too small to be tested properly.  Next I will attempt to increase the validation set slightly to see if that helps.  But first a test with 5 fold cross validation

Since all the classifiers are doing perfectly, I decided to try 5 fold cross-validation as something that could be used as a tie-breaker.  

In [45]:
from sklearn.model_selection import cross_val_score
Rand_For_Class.fit(X_train, y_train.values.ravel())
#cv = 5 means crossfold validation will be set to 5
scores_RandForest = cross_val_score(Rand_For_Class, X, y.values.ravel(), cv=5, scoring='accuracy')
#The score is shown after each fold
print(scores_RandForest)
print(scores_RandForest.mean())

[0.84255843 1.         0.99446154 1.         0.82142857]
0.9316897070948731


In [46]:
knn_LF = KNeighborsClassifier(n_neighbors = 5, metric = 'euclidean')
scores_knn = cross_val_score(knn_LF, X, y.values.ravel(), cv=5, scoring='accuracy')
print(scores_knn)
print(scores_knn.mean())

[0.84563346 1.         0.99569231 1.         0.68103448]
0.9044720493570984


In [47]:
svclassifier_LF = SVC(kernel = 'linear')
scores_SVC = cross_val_score(svclassifier_LF, X, y.values.ravel(), cv=5, scoring='accuracy')
print(scores_SVC)
print(scores_SVC.mean())

[0.84255843 1.         1.         1.         0.74753695]
0.9180190742794128


In [48]:
DTC_LF = DecisionTreeClassifier(random_state = 15)
scores_DTC = cross_val_score(DTC, X, y.values.ravel(), cv=5, scoring='accuracy')
print(scores_DTC)
print(scores_DTC.mean())

[0.84255843 1.         1.         1.         0.83990148]
0.9364919806833536


After everything it seems that either the Random Forest Classifier or the Decision tree classifiers would be a good bet.  Seeing as how they get a perfect score in the validation set, picking either the RandomForest or DecisionTree is acceptbale.  SCV and KNN were not far behind and would also be usable if need be. 

Now we can try with a slightly higher validation set

In [49]:
#Validation Set of .2

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=15)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.22, random_state=15)

# .7 Training / .2 val / .1 Test
# .9 of the original train * .22 will equal .198  Subtract .9 with .198 because ~.2 is the validation set and .7 is the train
#leaving the remaining .1 as the test set

In [50]:
Rand_For_Class.fit(X_train, y_train.values.ravel())
Forest_Pred_val2 = Rand_For_Class.predict(X_val)
target_names = ["Edible", 'Poisonous']

print(classification_report(y_val, Forest_Pred_val2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Forest_Pred_val2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy Score 1.0


In [51]:
tn, fp, fn, tp = confusion_matrix(y_val, Forest_Pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  851
False Positives:  0
False Negatives:  0
True Positives:  758


In [52]:
knn_val2 = KNeighborsClassifier(n_neighbors = 5, metric = 'euclidean')
knn_val2.fit(X_train, y_train.values.ravel())
Knn_Pred_val2 = knn_val2.predict(X_val)
print(classification_report(y_val, Knn_Pred_val2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_val2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy Score 1.0


In [53]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  851
False Positives:  0
False Negatives:  0
True Positives:  758


In [54]:
knn_val2 = KNeighborsClassifier(n_neighbors = 10, metric = 'euclidean')
knn_val2.fit(X_train, y_train.values.ravel())
Knn_Pred_val2 = knn_val2.predict(X_val)
print(classification_report(y_val, Knn_Pred_val2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_val2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy Score 0.9981354878806712


In [55]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  851
False Positives:  0
False Negatives:  3
True Positives:  755


In [56]:
knn_val2 = KNeighborsClassifier(n_neighbors = 10, metric = 'euclidean')
knn_val2.fit(X_train, y_train.values.ravel())
Knn_Pred_val2 = knn_val2.predict(X_val)
print(classification_report(y_val, Knn_Pred_val2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_val2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy Score 0.9981354878806712


In [57]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  851
False Positives:  0
False Negatives:  3
True Positives:  755


False Negative rate = False negative / True Positive + False negative = 3 / 755+3 = .004 = .4% chance of getting a mushroom was claimed to be edible but is actually poisonous.

In [58]:
knn_val2 = KNeighborsClassifier(n_neighbors = 20, metric = 'euclidean')
knn_val2.fit(X_train, y_train.values.ravel())
Knn_Pred_val2 = knn_val2.predict(X_val)
print(classification_report(y_val, Knn_Pred_val2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_val2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy Score 0.9981354878806712


In [59]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  851
False Positives:  0
False Negatives:  3
True Positives:  755


In [60]:
knn_val2 = KNeighborsClassifier(n_neighbors = 40, metric = 'euclidean')
knn_val2.fit(X_train, y_train.values.ravel())
Knn_Pred_val2 = knn_val2.predict(X_val)
print(classification_report(y_val, Knn_Pred_val2,target_names=target_names))
print("Accuracy Score " + str(accuracy_score(y_val, Knn_Pred_val2)))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy Score 0.997513983840895


In [61]:
tn, fp, fn, tp = confusion_matrix(y_val, Knn_Pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  850
False Positives:  1
False Negatives:  3
True Positives:  755


Here we get a mushroom which was labeled poisonous even though it was edible, which is not the worst thing, the three false edibles are the issue.

In [62]:
#SVM
svclassifier_val2 = SVC(kernel = 'linear')
svclassifier_val2.fit(X_train, y_train.values.ravel())

SVC_pred_val2 = svclassifier_val2.predict(X_val)

print(classification_report(y_val, SVC_pred_val2, target_names=target_names))
print("Accuracy:",metrics.accuracy_score(y_val, SVC_pred_val2))

              precision    recall  f1-score   support

      Edible       1.00      1.00      1.00       851
   Poisonous       1.00      1.00      1.00       758

   micro avg       1.00      1.00      1.00      1609
   macro avg       1.00      1.00      1.00      1609
weighted avg       1.00      1.00      1.00      1609

Accuracy: 1.0


In [63]:
tn, fp, fn, tp = confusion_matrix(y_val, SVC_pred_val2).ravel()
print("True Negatives: ",tn)
print("False Positives: ",fp)
print("False Negatives: ",fn)
print("True Positives: ",tp)

True Negatives:  851
False Positives:  0
False Negatives:  0
True Positives:  758


Given the data, it appears it is very good at seperating edible and poisonous mushrooms.  Our classifiers are almost always predicting a perfect score.  Meaning we can feel quite confident in eating our edible mushroom.  I would probably go in this order for our classifiers, RandomForest, DecisionTree, SVM, KNN.  That ordering is mostly based off of the cross validation but each classifier works very well.

Lets keep going though.  Lets see what features were so good at helping our classifiers label our mushrooms so accurately.  

In [64]:
#Random Forest classifier from our 20% validation set
feature_importances = pd.DataFrame(Rand_For_Class.feature_importances_, index = X_train.columns, columns=['importance']).sort_values('importance', ascending=False)

In [65]:
feature_importances.nlargest(10, 'importance')

importance
odor_n                        0.138694
odor_f                        0.135475
gill-size_n                   0.076017
stalk-surface-below-ring_k    0.070577
ring-type_p                   0.040980
population_v                  0.040896
spore-print-color_w           0.039499
bruises_t                     0.035984
spore-print-color_h           0.034039
ring-type_l                   0.029705

odor_n = none (no ordor)
odor_f = foul
gill-size_n = narrow
stalk-surface-below-ring_k = silky
ring-type_p = pendant
population_v = several
spore-print-color_w = white
bruises_t = True - There are bruises on the mushroom
spore-print-color_h = chocolate
rint-type_l = large

In [66]:
Dum  = pd.get_dummies(Shroom_DF)

In [67]:
Dum.corr()['class_e'].nlargest(10)

class_e                       1.000000
odor_n                        0.785557
ring-type_p                   0.540469
gill-size_b                   0.540024
bruises_t                     0.501530
stalk-surface-above-ring_s    0.491314
stalk-surface-below-ring_s    0.425444
spore-print-color_n           0.416645
spore-print-color_k           0.396832
gill-spacing_w                0.348387
Name: class_e, dtype: float64

In [68]:
Dum.corr()['class_e'].nsmallest(10)

class_p                      -1.000000
odor_f                       -0.623842
stalk-surface-above-ring_k   -0.587658
stalk-surface-below-ring_k   -0.573524
gill-size_n                  -0.540024
gill-color_b                 -0.538808
bruises_f                    -0.501530
spore-print-color_h          -0.490229
ring-type_l                  -0.451619
population_v                 -0.443722
Name: class_e, dtype: float64

In [69]:
Dum.corr()['class_p'].nlargest(10)

class_p                       1.000000
odor_f                        0.623842
stalk-surface-above-ring_k    0.587658
stalk-surface-below-ring_k    0.573524
gill-size_n                   0.540024
gill-color_b                  0.538808
bruises_f                     0.501530
spore-print-color_h           0.490229
ring-type_l                   0.451619
population_v                  0.443722
Name: class_p, dtype: float64

In [70]:
Dum.corr()['class_p'].nsmallest(10)

class_e                      -1.000000
odor_n                       -0.785557
ring-type_p                  -0.540469
gill-size_b                  -0.540024
bruises_t                    -0.501530
stalk-surface-above-ring_s   -0.491314
stalk-surface-below-ring_s   -0.425444
spore-print-color_n          -0.416645
spore-print-color_k          -0.396832
gill-spacing_w               -0.348387
Name: class_p, dtype: float64

These seem to make sense.  The top 10 important features are also the most correlated for either the edible or poisnonous mushrooms.  This means both have specific features that heavily impact the classifiers choice.  

odor_n = none (no ordor) - #1 for edible
odor_f = foul - #1 for poisonous
gill-size_n = narrow - #4 on poisonous
stalk-surface-below-ring_k = silky = #3 on poisonous 
ring-type_p = pendant #2 on edible
population_v = several #9 on poisonous
spore-print-color_w = white - not correlated well with either
bruises_t = True - There are bruises on the mushroom - #4 on edible
spore-print-color_h = chocolate #7 on poisonous
ring-type_l = large = #8 on poisonous


#Sources -
FeatureImportance - https://towardsdatascience.com/running-random-forests-inspect-the-feature-importances-with-this-code-2b00dd72b92e
TF/FN prints - https://towardsdatascience.com/demystifying-confusion-matrix-confusion-9e82201592fd
Previous assignments 